Ck tile batched gemm example

aledudek opened the pull request.

aledudek committed:
[CK Tile] Batched GEMM Example

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example

bartekxk commented on example/ck_tile/12_batched_gemm/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+add_executable(tile_example_batched_gemm EXCLUDE_FROM_ALL batched_gemm.cpp)
```
please add new line at the end

bartekxk commented on example/ck_tile/12_batched_gemm/run_batched_gemm_example.inc:
```diff
@@ -0,0 +1,214 @@
+    std::string op_name{"Gemm{MemBoundPipeline}"};
+
+    std::size_t flop = std::size_t(2) * M * N * K;
+    std::size_t num_byte =
```
please update num_byte and flop formulas. You have to include batch count

bartekxk commented on example/ck_tile/12_batched_gemm/run_batched_gemm_example.inc:
```diff
@@ -0,0 +1,214 @@
+                                       auto layout) {
+        if constexpr(std::is_same_v<decltype(layout), ck_tile::tensor_layout::gemm::RowMajor>)
+        {
+            return ck_tile::HostTensorDescriptor({batch_count_, row, col},
```
maybe replace row * col with batch_stride

bartekxk commented on include/ck_tile/host/reference/reference_gemm.hpp:
```diff
@@ -183,4 +183,116 @@ void reference_gemm_gpu(DeviceMem& a_device,
+    int numThreadsPerBlock = 256; // Common choice for threads per block
+    int numBlocks          = (totalElements + numThreadsPerBlock - 1) / numThreadsPerBlock;
+
+    for(int i = 0; i < batch_count; ++i)
```
maybe change `int i` to `index_t batch_id`

bartekxk commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2024, Advanced Micro Devices, Inc. All rights reserved.
```
```suggestion
// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
```

bartekxk commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,213 @@
+
+    using ADataType = remove_cvref_t<typename GemmPipeline::ADataType>;
+    using BDataType = remove_cvref_t<typename GemmPipeline::BDataType>;
+    // using CAccDataType = remove_cvref_t<typename GemmPipeline::CDataType>;
```
```suggestion
```

bartekxk commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,213 @@
+    CK_TILE_DEVICE void operator()(BatchedGemmCommonKargs kargs) const
+    {
+        const auto [i_m, i_n] = TilePartitioner{}();
+        const auto i_k        = blockIdx.z;
```
please change `i_k` to `i_batch`

bartekxk commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,213 @@
+    // using CAccDataType = remove_cvref_t<typename GemmPipeline::CDataType>;
+    using CDataType = remove_cvref_t<typename EpiloguePipeline::ODataType>;
+
+    __host__ static constexpr auto GridSize(index_t M, index_t N, index_t KBatch)
```
Could you add fourth parameter `BatchCount`? Then you can multiply KBatch times BatchCount. Add please add some check like this:
`if (KBatch > 1){ std::runtime_error("split k is not supported");}`

carlushuang commented on include/ck_tile/host/reference/reference_gemm.hpp:
```diff
@@ -183,4 +183,116 @@ void reference_gemm_gpu(DeviceMem& a_device,
+    BDataType* d_B;
+    CDataType* d_C;
+
+    hipError_t errA = hipMalloc(&d_A, batch_count * M * K * sizeof(ADataType));
```
I saw hipmalloc/memcpy inside gpu reference gemm. Can we remove this and allow outside caller to manage  the gpu buffer?
The reason is this function could be a fallback gpu function in the future, so we could have no malloc/memcpy inside this function (usually gpu memory management is by torch)

aledudek replied:
Yes, I will apply this change in a separate PR

aledudek committed:
[CK Tile] Batched GEMM Example - minor refactor

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example

aledudek committed:
[CK Tile] Batched GEMM Example - README update

bartekxk reviewed (dismissed):
LGTM

aosewski commented on example/ck_tile/13_batched_gemm/README.md:
````diff
@@ -0,0 +1,35 @@
+## example
+```
+args:
+        -k_batch     Batch Size (default:1)
````
We currently do not support Splitk so please remove this.

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.cpp:
```diff
@@ -0,0 +1,118 @@
+    auto kargs = Kernel::MakeKargs(args.p_a,
+                                   args.p_b,
+                                   args.p_c,
+                                   args.k_batch,
```
We do not yet support splitkK so please do not expose this parameter.

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.cpp:
```diff
@@ -0,0 +1,118 @@
+                                   args.batch_stride_C,
+                                   args.batch_count);
+
+    const dim3 grids      = Kernel::GridSize(args.M, args.N, args.k_batch, args.batch_count);
```
We do not yet support splitkK so please do not use k_batch parameter.

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.hpp:
```diff
@@ -0,0 +1,96 @@
+    using BDataType   = ck_tile::half_t;
+    using AccDataType = float;
+    using CDataType   = ck_tile::half_t;
+    // ToDo: Add more bias config to support different categories of GEMM.
```
Please remove this comment.

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.hpp:
```diff
@@ -0,0 +1,96 @@
+struct DataTypeTraits<ck_tile::half_t>
+{
+    static constexpr const char* name = "fp16";
+};
```
Are you using this? This is the same as for other gemm examples. Please do not duplicate.

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.hpp:
```diff
@@ -0,0 +1,96 @@
+    const void* p_a;
+    const void* p_b;
+    void* p_c;
+    ck_tile::index_t k_batch;
```
We do not support SplitK yet, thus please remove this param.

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.hpp:
```diff
@@ -0,0 +1,96 @@
+using AccDataType = Types::AccDataType;
+using CDataType   = Types::CDataType;
+
+struct batched_gemm_args
```
Take a look here: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/ops/topk_softmax/kernel/topk_softmax_kernel.hpp#L15-L25
I think that this is a good idea to place host args in kernel file.

aosewski replied:
Then here you just do like this: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/09_topk_softmax/topk_softmax_api.hpp#L17-L19

aosewski commented on example/ck_tile/13_batched_gemm/batched_gemm.hpp:
```diff
@@ -0,0 +1,96 @@
+auto create_args(int argc, char* argv[])
+{
+    ck_tile::ArgParser arg_parser;
+    arg_parser.insert("k_batch", "1", "Batch Size")
```
We need also layout for each input tensor.

aosewski commented on example/ck_tile/13_batched_gemm/run_batched_gemm_example.inc:
```diff
@@ -0,0 +1,220 @@
+
+    using ALayout = ck_tile::tensor_layout::gemm::RowMajor;
+    using BLayout = ck_tile::tensor_layout::gemm::ColumnMajor;
+    using CLayout = ck_tile::tensor_layout::gemm::RowMajor;
```
The layout could be defined in run time by the user. Take a look in here: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/03_gemm/run_gemm_example.inc

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+
+#include "ck_tile/core.hpp"
+#include "ck_tile/ops/common.hpp"
+#include "ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_scheduler.hpp"
```
Rather unused include.

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+        if(KBatch > 1)
+        {
+            std::runtime_error("split k is not supported");
+        }
```
I'd rather go with not exposing params which we do not support yet.

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+
+    __host__ static constexpr auto BlockSize() { return dim3(KernelBlockSize); }
+
+    struct BatchedGemmCommonKargs
```
```suggestion
    struct BatchedGemmKargs
```

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+        const auto i_batch    = blockIdx.z;
+        //  options
+        const ADataType* a_start = static_cast<const ADataType*>(kargs.a_ptr) +
+                                   __builtin_amdgcn_readfirstlane(i_batch * kargs.batch_stride_A);
```
This readfirstlane should be "outside" that is cover result of a_ptr + ibatch* stride

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+                    a_start,
+                    make_tuple(kargs.M, kargs.K),
+                    make_tuple(1, kargs.stride_A),
+                    number<GemmPipeline::VectorSizeA>{},
```
This is a bug, you can't use vector read on a not contiguous dimension. Take a look in here: https://github.com/ROCm/composable_kernel/blob/develop/include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp#L77-L95

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+                    b_start,
+                    make_tuple(kargs.N, kargs.K),
+                    make_tuple(kargs.stride_B, 1),
+                    number<GemmPipeline::VectorSizeB>{},
```
As above

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+        auto c_block_tile =
+            GemmPipeline{}.template operator()(a_block_window, b_block_window, num_loop, smem_ptr);
+
+        CDataType* c_start = static_cast<CDataType*>(kargs.c_ptr) +
+                             __builtin_amdgcn_readfirstlane(i_batch * kargs.batch_stride_C);
```
as above

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,220 @@
+                    c_start,
+                    make_tuple(kargs.M, kargs.N),
+                    make_tuple(1, kargs.stride_C),
+                    number<GemmPipeline::VectorSizeC>{},
```
As above

aosewski reviewed (changes requested):
Good start! But need to polish code a bit. Please also write unit tests.

aledudek committed:
[CK Tile] Batched Gemm Example - review changes

- Added tensor data layours as input parameters
- Changed structure of Host and Kernel args
- Removed bug with invalid vector read on non-contiguous memory

aledudek committed:
[CK Tile] Batched Gemm Example - remove comment

aledudek committed:
[CK Tile] Batched Gemm Example - Add GTests part1

aledudek committed:
[CK Tile] Batched Gemm Example - GTests part2 + review changes

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example

aledudek committed:
[CK TILE] Batched GEMM post merge fixes

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example

aosewski commented on example/ck_tile/16_batched_gemm/batched_gemm.hpp:
```diff
@@ -0,0 +1,63 @@
+}
+
+// host API
+float gemm_calc(batched_gemm_kargs args, const ck_tile::stream_config& s);
```
```suggestion
float batched_gemm(batched_gemm_kargs args, const ck_tile::stream_config& s);
```

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,227 @@
+            }
+        }();
+
+        auto a_pad_view = pad_tensor_view(
+            a_tensor_view,
+            make_tuple(number<TilePartitioner::kM>{}, number<TilePartitioner::kK>{}),
+            // somehow clang-format is splitting below line into multiple.
+            // clang-format off
+            sequence<false, GemmPipeline::kPadM>{});
```
Take a look at gemm_kernel how its done righ tnow.

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,227 @@
+            make_tuple(number<TilePartitioner::kM>{}, number<TilePartitioner::kK>{}),
+            {i_m, 0});
+
+        auto b_pad_view = pad_tensor_view(
```
Take a look at gemm_kernel how its done righ tnow.

aosewski commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -0,0 +1,227 @@
+            }
+        }();
+
+        auto c_pad_view = pad_tensor_view(
```
Take a look at gemm_kernel how its done righ tnow.

aosewski reviewed (changes requested):
Please catch-up with latest gemm-kernel.

aledudek committed:
[CK Tile] Batched GEMM Example - fix pad views

aosewski commented:
Let's wait for CI to pass.

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example

aledudek committed:
Merge branch 'develop' into ck_tile_batched_gemm_example
